### Background

You are the new Chief Data Scientist for your city's school district. In this capacity, you'll be helping the school board and mayor make strategic decisions regarding future school budgets and priorities.

As a first task, you've been asked to analyze the district-wide standardized test results. You'll be given access to every student's math and reading scores, as well as various information on the schools they attend. Your task is to aggregate the data to showcase obvious trends in school performance.


### Import Dependencies and Setup

In [182]:
# Dependencies and Setup
import pandas as pd

### Load, Read and Merge datasets

In [237]:
#Load and read csv file
schools_data_df = pd.read_csv('Resources/schools_complete.csv')
students_data_df = pd.read_csv('Resources/students_complete.csv')

schools_data_df.head()
students_data_df.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

### Datasets description

In [238]:
schools_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   School ID    15 non-null     int64 
 1   school_name  15 non-null     object
 2   type         15 non-null     object
 3   size         15 non-null     int64 
 4   budget       15 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 728.0+ bytes


In [239]:
schools_data_df.describe()

School ID         size        budget
count  15.000000    15.000000  1.500000e+01
mean    7.000000  2611.333333  1.643295e+06
std     4.472136  1420.915282  9.347763e+05
min     0.000000   427.000000  2.480870e+05
25%     3.500000  1698.000000  1.046265e+06
50%     7.000000  2283.000000  1.319574e+06
75%    10.500000  3474.000000  2.228999e+06
max    14.000000  4976.000000  3.124928e+06

In [240]:
students_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39170 entries, 0 to 39169
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Student ID     39170 non-null  int64 
 1   student_name   39170 non-null  object
 2   gender         39170 non-null  object
 3   grade          39170 non-null  object
 4   school_name    39170 non-null  object
 5   reading_score  39170 non-null  int64 
 6   math_score     39170 non-null  int64 
dtypes: int64(3), object(4)
memory usage: 2.1+ MB


In [241]:
students_data_df.describe()

Student ID  reading_score    math_score
count  39170.000000    39170.00000  39170.000000
mean   19584.500000       81.87784     78.985371
std    11307.549359       10.23958     12.309968
min        0.000000       63.00000     55.000000
25%     9792.250000       73.00000     69.000000
50%    19584.500000       82.00000     79.000000
75%    29376.750000       91.00000     89.000000
max    39169.000000       99.00000     99.000000

### Distirct Summary
Perform the necessary calculations and then create a high-level snapshot of the district's key metrics in a DataFrame.

Include the following:

- Total number of unique schools

- Total students

- Total budget

- Average math score

- Average reading score

- % passing math (the percentage of students who passed math)

- % passing reading (the percentage of students who passed reading)

- % overall passing (the percentage of students who passed math AND reading)

In [242]:
#combine th 2 dataset into one dataset
district_summary = pd.merge(students_data_df, schools_data_df)
district_summary.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

In [243]:
#calculate the total number of unqiue schools
school_counts = district_summary['school_name'].nunique()
school_counts

15

In [244]:
# Total students
student_counts = district_summary['student_name'].count()
student_counts

39170

In [245]:
# Total budget
total_budget = schools_data_df['budget'].sum()
total_budget

24649428

In [246]:
# calculate the average math score
avg_math_score = district_summary['math_score'].mean()
avg_math_score

78.98537145774827

In [247]:
# calculate the average reading score
avg_reading_score = district_summary['reading_score'].mean()
avg_reading_score


81.87784018381414

In [248]:
passing_math_count = district_summary[(district_summary['math_score'] >= 70)].count()['student_name']
passing_math_percentage = passing_math_count / float(student_counts) * 100
passing_math_percentage

74.9808526933878

In [249]:
passing_read_count = district_summary[(district_summary['reading_score'] >= 70)].count()['student_name']
passing_read_percentage = passing_read_count / float(student_counts) * 100
passing_read_percentage

85.80546336482001

In [250]:
overall_passing_count = district_summary[(district_summary['math_score']>=70) & (district_summary['reading_score']>=70)].count()['student_name']
overall_passing_rate = overall_passing_count/float(student_counts)*100
overall_passing_rate

65.17232575950983

In [252]:
summary_district = pd.DataFrame({
    'Total schools': school_counts, 
    'Total students':f"{student_counts:,}", 
    'Total budget': f"${total_budget:,.2f}",
    'Average math score': f"{avg_math_score:.6f}",
    'Average read score': f"{avg_reading_score:.5f}",
    '% passing math': f"{passing_math_percentage:.6f}",
    '% passing reading': f"{passing_read_percentage:.6f}",
    '% overall passing': f"{overall_passing_rate:.6f}"}, 
    index=[0])
summary_district


Total schools Total students    Total budget Average math score  \
0             15         39,170  $24,649,428.00          78.985371   

  Average read score % passing math % passing reading % overall passing  
0           81.87784      74.980853         85.805463         65.172326

### School Summary

Perform the necessary calculations and then create a DataFrame that summarizes key metrics about each school.

Include the following:

- School name

- School type

- Total students

- Total school budget

- Per student budget

- Average math score

- Average reading score

- % passing math (the percentage of students who passed math)

- % passing reading (the percentage of students who passed reading)

- % overall passing (the percentage of students who passed math AND reading)

In [198]:
# School name
school_name = district_summary.set_index('school_name').groupby(['school_name'])
school_name

In [199]:
# School Type
school_type = schools_data_df.set_index('school_name')['type']
school_type

school_name
Huang High School        District
Figueroa High School     District
Shelton High School       Charter
Hernandez High School    District
Griffin High School       Charter
Wilson High School        Charter
Cabrera High School       Charter
Bailey High School       District
Holden High School        Charter
Pena High School          Charter
Wright High School        Charter
Rodriguez High School    District
Johnson High School      District
Ford High School         District
Thomas High School        Charter
Name: type, dtype: object

In [200]:
# Total students count using student id
school_students_count = school_name['Student ID'].count()
school_students_count

school_name
Bailey High School       4976
Cabrera High School      1858
Figueroa High School     2949
Ford High School         2739
Griffin High School      1468
Hernandez High School    4635
Holden High School        427
Huang High School        2917
Johnson High School      4761
Pena High School          962
Rodriguez High School    3999
Shelton High School      1761
Thomas High School       1635
Wilson High School       2283
Wright High School       1800
Name: Student ID, dtype: int64

In [201]:
# total_budget
total_school_budget = district_summary.groupby(["school_name"]).mean()["budget"]
total_school_budget

school_name
Bailey High School       3124928.0
Cabrera High School      1081356.0
Figueroa High School     1884411.0
Ford High School         1763916.0
Griffin High School       917500.0
Hernandez High School    3022020.0
Holden High School        248087.0
Huang High School        1910635.0
Johnson High School      3094650.0
Pena High School          585858.0
Rodriguez High School    2547363.0
Shelton High School      1056600.0
Thomas High School       1043130.0
Wilson High School       1319574.0
Wright High School       1049400.0
Name: budget, dtype: float64

In [202]:
# Per student budget
per_student_budget = total_school_budget/school_students_count
per_student_budget

school_name
Bailey High School       628.0
Cabrera High School      582.0
Figueroa High School     639.0
Ford High School         644.0
Griffin High School      625.0
Hernandez High School    652.0
Holden High School       581.0
Huang High School        655.0
Johnson High School      650.0
Pena High School         609.0
Rodriguez High School    637.0
Shelton High School      600.0
Thomas High School       638.0
Wilson High School       578.0
Wright High School       583.0
dtype: float64

In [203]:
# Average math score
avg_math_score = school_name['math_score'].mean()
avg_math_score

school_name
Bailey High School       77.048432
Cabrera High School      83.061895
Figueroa High School     76.711767
Ford High School         77.102592
Griffin High School      83.351499
Hernandez High School    77.289752
Holden High School       83.803279
Huang High School        76.629414
Johnson High School      77.072464
Pena High School         83.839917
Rodriguez High School    76.842711
Shelton High School      83.359455
Thomas High School       83.418349
Wilson High School       83.274201
Wright High School       83.682222
Name: math_score, dtype: float64

In [204]:
# Average reading score
avg_reading_score = school_name['reading_score'].mean()
avg_reading_score

school_name
Bailey High School       81.033963
Cabrera High School      83.975780
Figueroa High School     81.158020
Ford High School         80.746258
Griffin High School      83.816757
Hernandez High School    80.934412
Holden High School       83.814988
Huang High School        81.182722
Johnson High School      80.966394
Pena High School         84.044699
Rodriguez High School    80.744686
Shelton High School      83.725724
Thomas High School       83.848930
Wilson High School       83.989488
Wright High School       83.955000
Name: reading_score, dtype: float64

In [205]:
# % passing math (the percentage of students who passed math)
passing_math_count = district_summary[(district_summary['math_score'] >= 70)].groupby('school_name')['Student ID'].count()
passing_math_percentage = passing_math_count / (school_students_count) * 100
passing_math_percentage

school_name
Bailey High School       66.680064
Cabrera High School      94.133477
Figueroa High School     65.988471
Ford High School         68.309602
Griffin High School      93.392371
Hernandez High School    66.752967
Holden High School       92.505855
Huang High School        65.683922
Johnson High School      66.057551
Pena High School         94.594595
Rodriguez High School    66.366592
Shelton High School      93.867121
Thomas High School       93.272171
Wilson High School       93.867718
Wright High School       93.333333
Name: Student ID, dtype: float64

In [206]:
# % passing reading (the percentage of students who passed reading)
passing_reading_count = district_summary[(district_summary['reading_score'] >= 70)].groupby('school_name')['Student ID'].count()
passing_reading_percentage = passing_reading_count / (school_students_count) * 100
passing_reading_percentage

school_name
Bailey High School       81.933280
Cabrera High School      97.039828
Figueroa High School     80.739234
Ford High School         79.299014
Griffin High School      97.138965
Hernandez High School    80.862999
Holden High School       96.252927
Huang High School        81.316421
Johnson High School      81.222432
Pena High School         95.945946
Rodriguez High School    80.220055
Shelton High School      95.854628
Thomas High School       97.308869
Wilson High School       96.539641
Wright High School       96.611111
Name: Student ID, dtype: float64

In [207]:
# % overall passing (the percentage of students who passed math AND reading)
overall_passing_count = district_summary[(district_summary['math_score'] >= 70) & (district_summary['reading_score']>69)].groupby('school_name')['Student ID'].count()
overall_passing_percentage = overall_passing_count / (school_students_count) * 100
overall_passing_percentage

school_name
Bailey High School       54.642283
Cabrera High School      91.334769
Figueroa High School     53.204476
Ford High School         54.289887
Griffin High School      90.599455
Hernandez High School    53.527508
Holden High School       89.227166
Huang High School        53.513884
Johnson High School      53.539172
Pena High School         90.540541
Rodriguez High School    52.988247
Shelton High School      89.892107
Thomas High School       90.948012
Wilson High School       90.582567
Wright High School       90.333333
Name: Student ID, dtype: float64

In [216]:
per_school_summary= pd.DataFrame({'School Type': school_type,
                                  'Total students': school_students_count,
                                  'Total school budget': total_school_budget,
                                  'Per student budget': per_student_budget, 
                                  'Average math score': avg_math_score,
                                  'Average reading score': avg_reading_score,
                                  '% passing math': passing_math_percentage,
                                  '% passing reading': passing_reading_percentage,
                                  '% overall passing': overall_passing_percentage})

per_school_summary = per_school_summary[['School Type','Total students',
                                        'Total school budget','Per student budget',
                                        'Average math score','Average reading score',
                                        '% passing math','% passing reading', '% overall passing']]
per_school_summary

School Type  Total students  Total school budget  \
Bailey High School       District            4976            3124928.0   
Cabrera High School       Charter            1858            1081356.0   
Figueroa High School     District            2949            1884411.0   
Ford High School         District            2739            1763916.0   
Griffin High School       Charter            1468             917500.0   
Hernandez High School    District            4635            3022020.0   
Holden High School        Charter             427             248087.0   
Huang High School        District            2917            1910635.0   
Johnson High School      District            4761            3094650.0   
Pena High School          Charter             962             585858.0   
Rodriguez High School    District            3999            2547363.0   
Shelton High School       Charter            1761            1056600.0   
Thomas High School        Charter            1635            1043130.0   
Wilson High School        Charter            2283            1319574.0   
Wright High School        Charter            1800            1049400.0   

                       Per student budget  Average math score  \
Bailey High School                  628.0           77.048432   
Cabrera High School                 582.0           83.061895   
Figueroa High School                639.0           76.711767   
Ford High School                    644.0           77.102592   
Griffin High School                 625.0           83.351499   
Hernandez High School               652.0           77.289752   
Holden High School                  581.0           83.803279   
Huang High School                   655.0           76.629414   
Johnson High School                 650.0           77.072464   
Pena High School                    609.0           83.839917   
Rodriguez High School               637.0           76.842711   
Shelton High School                 600.0           83.359455   
Thomas High School                  638.0           83.418349   
Wilson High School                  578.0           83.274201   
Wright High School                  583.0           83.682222   

                       Average reading score  % passing math  \
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % passing reading  % overall passing  
Bailey High School             81.933280          54.642283  
Cabrera High School            97.039828          91.334769  
Figueroa High School           80.739234          53.204476  
Ford High School               79.299014          54.289887  
Griffin High School            97.138965          90.599455  
Hernandez High School          80.862999          53.527508  
Holden High School             96.252927          89.227166  
Huang High School              81.316421          53.513884  
Johnson High School            81.222432          53.539172  
Pena High School               95.945946          90.540541  
Rodriguez High School          80.220055          52.988247  
Shelton High School        

### Highest-Performing Schools by Percentage of Overall Passing
- Sort the schools by % Overall Passing in descending order
- Save the results to a DataFrame called top_schools 
- Display the first 5 rows 

In [209]:
highest_performing = per_school_summary.sort_values('% overall passing', ascending = False)
top_schools = pd.DataFrame(highest_performing)
top_schools.head()

School Type  Total students  Total school budget  \
school_name                                                            
Cabrera High School     Charter            1858            1081356.0   
Thomas High School      Charter            1635            1043130.0   
Griffin High School     Charter            1468             917500.0   
Wilson High School      Charter            2283            1319574.0   
Pena High School        Charter             962             585858.0   

                     Per student budget  Average math score  \
school_name                                                   
Cabrera High School               582.0           83.061895   
Thomas High School                638.0           83.418349   
Griffin High School               625.0           83.351499   
Wilson High School                578.0           83.274201   
Pena High School                  609.0           83.839917   

                     Average reading score  % passing math  % passing reading  \
school_name                                                                     
Cabrera High School              83.975780       94.133477          97.039828   
Thomas High School               83.848930       93.272171          97.308869   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   
Pena High School                 84.044699       94.594595          95.945946   

                     % overall passing  
school_name                             
Cabrera High School          91.334769  
Thomas High School           90.948012  
Griffin High School          90.599455  
Wilson High School           90.582567  
Pena High School             90.540541

### Lowest-Performing Schools by Percentage of Overall Passing 
- Sort the schools by % Overall Passing in ascending order
- Save the results to a DataFrame called bottom_schools
- Display the first 5 rows

In [210]:
lowest_performing = per_school_summary.sort_values('% overall passing')
bottom_schools = pd.DataFrame(lowest_performing)
bottom_schools.head()

School Type  Total students  Total school budget  \
school_name                                                              
Rodriguez High School    District            3999            2547363.0   
Figueroa High School     District            2949            1884411.0   
Huang High School        District            2917            1910635.0   
Hernandez High School    District            4635            3022020.0   
Johnson High School      District            4761            3094650.0   

                       Per student budget  Average math score  \
school_name                                                     
Rodriguez High School               637.0           76.842711   
Figueroa High School                639.0           76.711767   
Huang High School                   655.0           76.629414   
Hernandez High School               652.0           77.289752   
Johnson High School                 650.0           77.072464   

                       Average reading score  % passing math  \
school_name                                                    
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Hernandez High School              80.934412       66.752967   
Johnson High School                80.966394       66.057551   

                       % passing reading  % overall passing  
school_name                                                  
Rodriguez High School          80.220055          52.988247  
Figueroa High School           80.739234          53.204476  
Huang High School              81.316421          53.513884  
Hernandez High School          80.862999          53.527508  
Johnson High School            81.222432          53.539172

### Math Scores by Grade 
Use the code provided to separate the data by grade
Group by "school_name" and take the mean of each 
Use the code to select only the math_score 
Combine each of the scores above into single DataFrame called math_scores_by_grade 

In [214]:
# Use the code provided to separate the data by grade
ninth_graders = district_summary[(district_summary["grade"] == "9th")]
tenth_graders = district_summary[(district_summary["grade"] == "10th")]
eleventh_graders = district_summary[(district_summary["grade"] == "11th")]
twelfth_graders = district_summary[(district_summary["grade"] == "12th")]

# Group by "school_name" and take the mean of each.
ninth_graders_scores = ninth_graders.groupby('school_name')['math_score'].mean()
tenth_graders_scores = tenth_graders.groupby('school_name')['math_score'].mean()
eleventh_graders_scores = eleventh_graders.groupby('school_name')['math_score'].mean() 
twelfth_graders_scores = twelfth_graders.groupby('school_name')['math_score'].mean()

# Creat the math scores data frame by grade level
math_scores_by_grade =pd.DataFrame({'9th': ninth_graders_scores, 
                           '10th': tenth_graders_scores,
                          '11th': eleventh_graders_scores,
                          '12th': twelfth_graders_scores})

math_scores_by_grade = math_scores_by_grade[['9th','10th','11th','12th']]
math_scores_by_grade.index.name = 'School Name'
# Display the DataFrame
math_scores_by_grade

9th       10th       11th       12th
School Name                                                      
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

### Reading Scores by Grade 
Use the code provided to separate the data by grade 
Group by "school_name" and take the mean of each 
Use the code to select only the reading_score 
Combine each of the scores above into single DataFrame called reading_scores_by_grade 

In [215]:
ninth_graders = district_summary[(district_summary["grade"] == "9th")]
tenth_graders = district_summary[(district_summary["grade"] == "10th")]
eleventh_graders = district_summary[(district_summary["grade"] == "11th")]
twelfth_graders = district_summary[(district_summary["grade"] == "12th")]

# Group by "school_name" and take the mean of each.
ninth_graders_scores = ninth_graders.groupby('school_name')['reading_score'].mean()
tenth_graders_scores = tenth_graders.groupby('school_name')['reading_score'].mean()
eleventh_graders_scores = eleventh_graders.groupby('school_name')['reading_score'].mean() 
twelfth_graders_scores = twelfth_graders.groupby('school_name')['reading_score'].mean()

# Creat the math scores data frame by grade level
reading_scores_by_grade =pd.DataFrame({'9th': ninth_graders_scores, 
                           '10th': tenth_graders_scores,
                          '11th': eleventh_graders_scores,
                          '12th': twelfth_graders_scores})

reading_scores_by_grade = reading_scores_by_grade[['9th','10th','11th','12th']]
reading_scores_by_grade.index.name = 'School Name'
# Display the DataFrame
reading_scores_by_grade

9th       10th       11th       12th
School Name                                                      
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

### Scores by School Spending 
Use pd.cut with the provided code to bin the data by the spending ranges 
Use the code provided to calculate the averages 
Create the spending_summary DataFrame using the binned and averaged spending data 

In [230]:
# Establish the bins 
spending_bins = [0, 585, 630, 645, 680]
labels = ["<$585", "$585-630", "$630-645", "$645-680"]

# Create a copy of the school summary since it has the "Per Student Budget" 
school_spending_df = per_school_summary.copy()

# Use `pd.cut` to categorize spending based on the bins.
school_spending_df["Spending Ranges (Per Student)"] = pd.cut(school_spending_df['Per student budget'], bins = spending_bins, labels = labels)

#group by spending
by_school_spending = school_spending_df.groupby('Spending Ranges (Per Student)')

spending_math_scores = by_school_spending.mean()["Average math score"]
spending_reading_scores = by_school_spending.mean()["Average reading score"]
spending_passing_math = by_school_spending.mean()["% passing math"]
spending_passing_reading = by_school_spending.mean()["% passing reading"]
overall_passing_spending = by_school_spending.mean()["% overall passing"]


# df build            
scores_by_spend = pd.DataFrame({
    "Average math score": spending_math_scores,
    "Average reading score": spending_reading_scores,
    "% passing math": spending_passing_math,
    "% passing reading": spending_passing_reading,
    "% overall passing": overall_passing_spending
            
})
            
#reorder columns
scores_by_spend = scores_by_spend[[
    "Average math score",
    "Average reading score",
    "% passing math",
    "% passing reading",
    "% overall passing"
]]

scores_by_spend.index.name = "Spending Ranges (Per Student)"

scores_by_spend









Average math score  Average reading score  \
Spending Ranges (Per Student)                                              
<$585                                   83.455399              83.933814   
$585-630                                81.899826              83.155286   
$630-645                                78.518855              81.624473   
$645-680                                76.997210              81.027843   

                               % passing math  % passing reading  \
Spending Ranges (Per Student)                                      
<$585                               93.460096          96.610877   
$585-630                            87.133538          92.718205   
$630-645                            73.484209          84.391793   
$645-680                            66.164813          81.133951   

                               % overall passing  
Spending Ranges (Per Student)                     
<$585                                  90.369459  
$585-630                               81.418596  
$630-645                               62.857656  
$645-680                               53.526855

### Scores by School Size 
Use pd.cut with the provided code to bin the data by the school sizes 
Use the code provided to calculate the averages 
Create the size_summary DataFrame using the binned and averaged size data 

In [233]:
# Establish the bins.
size_bins = [0, 1000, 2000, 5000]
labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]
# Categorize the spending based on the bins
# Use `pd.cut` on the "Total Students" column of the `per_school_summary` DataFrame.
# referencing the school_spending_df dateframe defined from score by spending calculations

per_school_summary["School Size"] =  pd.cut(per_school_summary['Total students'], bins = size_bins, labels = labels)

#group by spending
by_school_size = per_school_summary.groupby('School Size')

size_math_scores = by_school_size.mean()["Average math score"]
size_reading_scores = by_school_size.mean()["Average reading score"]
size_passing_math = by_school_size.mean()["% passing math"]
size_passing_reading = by_school_size.mean()["% passing reading"]
size_overall_passing = by_school_size.mean()["% overall passing"]


# df build            
scores_by_size = pd.DataFrame({
    "Average math score": size_math_scores,
    "Average reading score": size_reading_scores,
    "% passing math": size_passing_math,
    "% passing reading": size_passing_reading,
    "% overall passing": size_overall_passing
            
})
            
#reorder columns
scores_by_size = scores_by_size[[
    "Average math score",
    "Average reading score",
    "% passing math",
    "% passing reading",
    "% overall passing"
]]

scores_by_size.index.name = "School Size"

scores_by_size





Average math score  Average reading score  % passing math  \
School Size                                                                     
Small (<1000)                83.821598              83.929843       93.550225   
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % passing reading  % overall passing  
School Size                                               
Small (<1000)               96.099437          89.883853  
Medium (1000-2000)          96.790680          90.621535  
Large (2000-5000)           82.766634          58.286003

### Scores by School Type
Group the per_school_summary DataFrame by "School Type" and average the results 
Use the code provided to select the new column data 
Create a new DataFrame called type_summary that uses the new column data 

In [234]:
#group by spending
by_school_type = per_school_summary.groupby('School Type')

type_math_scores = by_school_type.mean()["Average math score"]
type_reading_scores = by_school_type.mean()["Average reading score"]
type_passing_math = by_school_type.mean()["% passing math"]
type_passing_reading = by_school_type.mean()["% passing reading"]
type_overall_passing = by_school_type.mean()["% overall passing"]


# df build            
scores_by_type = pd.DataFrame({
    "Average math score": type_math_scores,
    "Average reading score": type_reading_scores,
    "% passing math": type_passing_math,
    "% passing reading": type_passing_reading,
    "% overall passing": type_overall_passing
            
})
            
#reorder columns
scores_by_type = scores_by_type[[
    "Average math score",
    "Average reading score",
    "% passing math",
    "% passing reading",
    "% overall passing"
]]

scores_by_type.index.name = "School Type"

scores_by_type



Average math score  Average reading score  % passing math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % passing reading  % overall passing  
School Type                                        
Charter              96.586489          90.432244  
District             80.799062          53.672208

### Written Report
To receive all points, the written report presents a cohesive written analysis that:

Summarizes the analysis 

Draws two correct conclusions or comparisons from the calculations 

### Observable Trends

- Two datasets were provided for this analysis, which are school data and student data. In combining this two datasets, it is determined there are 15 schools and 39,170 students.

- The total students average reading score is 81.87 which is higher than the average math score 78.98. And the passing reading rate of 85.80 is higher than passing math rate of 74.98 %, with the overall passing rate is 65.17%. Based on this findings, it can be inferred that student have greater competency in reading than math. 

- The school summary showed that 7 out of the total 15 schools are at the district level and other 8 eight at charter level. Also, it is indicated from the analysis that distrcit schools have more students an budget than charter school due geographical coverage.

- The total budget field of the school summery showed the budget amount allocated to each school. Bailey high school has the  maximum budget of 3,124,928.00 and Holden High school has the lowest budget of 248,087.00 its Holden High School. We can see later Bailey High School is not in the top performing schools. Regarding budget per student, Huang High School (655.000) has a higher budget per student, and Wilson High School (578.00) allocates a lower budget per student. Huang High School is one of the bottom performing school. The total district schools budget( 17,347,923.00) is higher than Charter schools budget (7,301,505.00), however, when we come to the total allocation of Budget Per Student charter schools allocation (4,796.00) is greater than district schools (4,505.00). This may due to charter schools have a smaller number of students.

- The findings from the data showed that in all metrics (average, percent, and overall) of reading and math scores Charter schools have a greater score than District schools. Pena High School has the maximum (83.839917) average math score (83.839917), percent of students passing math (94.59%) and average reading score (84.044699). it can be inferred that Pena High School might have a good quality, and reputation in teaching math, and reading; further investigation on the contributing factors are required to support the hypothesis. On the other side, Huang High School has a lower average math score (76.842711), and percent of students passing math (65.683922). Regarding the percent of passing reading class Thomas High School has a higher percentage value (97.30%). On the flip side, Rodriguez High School has a lower Average Reading Score(80.744686), and Ford High School has a lower passing reading. In the overall percentage passing Cabrera High School has the maximum passing percentage rate ( 91.33 %), and Rodriguez High School has a lower percentage of (52.98) passing.

- School by school spending analysis does show inverse relationship between the spending range(per student) and overall grade scores for students in math and reading. The same inverse relationship is observed in the "school by school sizes analysis". The overall grade score for math and reading category decreases with increase in student size.

- it is determined that charter level school with the least budget and student size performed better than district level schools with bigger budget and student size.

- Grade 11 students at Holden High School holds a higher average Math score (85) comparing all schools and grades. Holden High School grade 12 students possess a higher average reading score (84.698795).

- The data revealed that when the number of students sizes higher in schools the average math scores, average reading scores, passing math, passing reading, and overall passing decrease. So, we can infer that, when student size body increase, it negatively influences student achievement. Therefore, schools, cities, districts, and charters should work on student size in schools, and classes because it directly affects student achievement for so many reasons.